topics/S3/etcd-migration/src/steps.py

```diff
@@ -0,0 +1,196 @@
+    if migration_status["error"] != "":
+        print(f"{bcolors.RED}ERROR: migration failed during step {migration_status['step']} with error: "
+              f"'{migration_status['error']}'. Fix what is needed and rerun script{bcolors.ENDC}")
+        exit(1)
+    print(f"{bcolors.GREEN}\tFinished migrating data.{bcolors.ENDC}")
+
+
+@step("Validating migration")
+def validate_migration():
+    data = {"kwas_endpoint": global_vars.socket_path}
+    r = send_request(f"http://{global_vars.s3_hosts[0]}:9001/minio/migration/validate/kwas", data).json()
+    if not r['verified']:
+        print(f"{bcolors.RED}ERROR: Migration Validation failed. Fix it and run script again. The invalid data is: {bcolors.ENDC}")
+        print_errors_dict(r["errors"])
+        exit(1)
+    print(f"{bcolors.GREEN}\tAll migrated data was verified successfully{bcolors.ENDC}")
+
+
+@step("Drain all S3 hosts and Restart them into new mode")
+def drain_and_restarts_all_hosts():
+    for host in global_vars.s3_hosts:
+        if global_vars.auto_mode < Automode.AUTO.value:
+            if not continue_to_next_host():
+                do_revert()
+
+        set_drain_mode(True, int(''.join(filter(str.isdigit, host))))
+        print(f"{bcolors.CYAN}\tWaiting for host {host} to be drained. if this takes too long, cancel "
+              f"script (^C) fix what is needed and re-run script{bcolors.ENDC}")
+        wait_for_drain(host)
+        print(f"{bcolors.DARK_GREEN}\tHost {host} is in Drain Mode, Restarting it into new mode{bcolors.ENDC}")
+
+        output = send_bash_command(f"weka debug manhole --slot 0 --host {int(''.join(filter(str.isdigit, host)))} "
+                                   f"s3_update_config force_minio_refresh=true")
+        while True:
+            time.sleep(2)
+            output = send_bash_command("weka s3 cluster status -J")
+            hosts_health = json.loads(output)
+            if all(val for val in hosts_health.values()):
+                break
+
+    print(f"{bcolors.GREEN}\tAll hosts are Ready.{bcolors.ENDC}")
+
+
+@step("Validate all minIOs are up with KWAS")
+def validate_kwas_is_up():
+    for host in global_vars.s3_hosts:
+        r = send_request(f"http://{host}:9001/minio/migration/clients")
+        response_body = r.json()
+        if not (response_body['kwas client'] == "true" and response_body['etcd client'] == "false"):
+            print(f"{bcolors.RED}ERROR: Not all minIOs are up with kwas. Fix it and run script again. The invalid data is: {bcolors.ENDC}")
+            exit(1)
+    print(f"{bcolors.GREEN}\tAll hosts are Ready with KWAS.{bcolors.ENDC}")
+
+    print(f"{bcolors.YELLOW}{bcolors.BOLD}After this step, rolling back to ETCD is no longer supported. "
+          f"Meaning this is the step after which the S3 cluster is forever KWAS.{bcolors.ENDC}")
+    global_vars.auto_mode = 1
+
+
+@step("Exit S3 cluster from Migration Read Only Mode")
+def exit_migration_mode():
+    for host in global_vars.s3_hosts:
```

Comment:
there's a lot of duplicated code between this and `enter_migratino_mode` function. maybe export the logic generically to an external function that both steps will use